### DISCUSSION TITLE
1.19.3 OreGen / ConfiguredFeature Usage

### DISCUSSION THREAD

**Discussion by rshchekotov** [FabricMC/fabric-api#2793] [Mod Dev Support] [answered]
**In short**: I was following a tutorial and implementing Ore Generation using Configured Features until I realized that it's done differently in 1.19.3 (since a few Constants/Methods went missing). I checked out the [1.19.3 update post](https://fabricmc.net/2022/11/24/1193.html#Custom-world-generation), and it briefly mentioned that the OreGen is now done using the DataGen API, as of my understanding. I searched the web and tried to implement it myself, but my Fabric API knowledge is limited (it's my second day working with it) and there seem to be no docs online regarding OreGen using DataGen (or more generally: Structures/Features using DataGen).

**More details** can be found in #2792 (Next time I will write up the post here directly :sweat_smile:, apologies).

Thanks in advance :smile:

**Comment by warjort** [comment, marked as answer]
I don't know if there are any tutorials for 1.19.3 yet?

You can find this example in the fabric api testsuite: 
https://github.com/FabricMC/fabric/blob/1.19.3/fabric-biome-api-v1/src/testmod/java/net/fabricmc/fabric/test/biome/WorldgenProvider.java

But you know you can just put your configured/placed features in json files in your src/main/resources/data
Your mod is a datapack so all the normal rules apply.

Docs for datapacks: https://minecraft.fandom.com/wiki/Data_pack

A"gui" to help you generate them:
https://misode.github.io/

Vanilla's iron generation as json files:
https://github.com/misode/mcmeta/blob/data/data/minecraft/worldgen/configured_feature/ore_iron.json
https://github.com/misode/mcmeta/blob/data/data/minecraft/worldgen/placed_feature/ore_iron_middle.json

The datagen just lets you generate these files in a program.

**Reply by rshchekotov** [reply]
Firstly, thanks a lot for the quick reply!
Yes, I am aware - I would much prefer the DataGen solution, as it would allow me to write a single line and generate a huge JSON out of it, opposed to copy-pasting the JSON every time I add a new ore/structure.

I appreciate the resources and input though!
I believe I can get on from here :D

**Comment by apple502j** [comment]
@rshchekotov FYI, your issue is probably that you passed the wrong registriesFuture in the provider constructor. You pass the argument as is, you don't make one yourself.

**Reply by rshchekotov** [reply]
Heh, I was **just** writing about this. Thanks **a lot** for noticing, I resolved it already :smile:

**Comment by rshchekotov** [comment]
Hey @warjort (and @apple502j), just wanted to thank you once again & let you know - I made it work!
The root-issue stemmed from my own oversight (that there's a `(output, future) -> {}`-factory and that I don't have to construct that future myself and also the fact that some things regarding registries changed a little compared to 1.18.2.

If someone will stumble upon this discussion in search for a solution, here's my way of doing it (It's Kotlin, but it should be easy to translate to Java if necessary, imports are not included, if not clear - feel free to ask!):

Assuming you have: `ModBlocks.EXAMPLE_BLOCK` & `ModBlocks.DEEPSLATE_EXAMPLE_BLOCK`

I prefer to split things in two, a "ModFeatures"-class and the "DataGen"-class, first the "ModFeatures":
```kt
object ModOres {
  val EXAMPLE_ORE = registerDualOre(
    ModBlocks.EXAMPLE_BLOCK, ModBlocks.DEEPSLATE_EXAMPLE_BLOCK,
    "example_block", 9)

  fun registerDualOre(block: Block, deepslate: Block, key: String, size: Int): OreGenRegistry {
    val featureConfig = OreFeatureConfig(
      listOf(
        OreFeatureConfig.createTarget(TagMatchRuleTest(BlockTags.STONE_ORE_REPLACEABLES), block.defaultState),
        OreFeatureConfig.createTarget(TagMatchRuleTest(BlockTags.DEEPSLATE_ORE_REPLACEABLES), deepslate.defaultState)
      ), size)
    val feature = ConfiguredFeature(Feature.ORE, featureConfig)
    val configured = RegistryKey.of(RegistryKeys.CONFIGURED_FEATURE, custom(key))
    val placed = RegistryKey.of(RegistryKeys.PLACED_FEATURE, custom(key))
    return OreGenEntry(feature, configured, placed)
  }

  data class OreGenRegistry(
    val feature: ConfiguredFeature<OreFeatureConfig, Feature<OreFeatureConfig>>,
    val configured: RegistryKey<ConfiguredFeature<*, *>>,
    val placed: RegistryKey<PlacedFeature>)
}
```

And then, the "registered" Ore can be data-gen'ed as follows:
```kt
class OreGenerationProvider(
    output: FabricDataOutput?,
    registriesFuture: CompletableFuture<RegistryWrapper.WrapperLookup>
) : FabricDynamicRegistryProvider(output, registriesFuture) {
    override fun getName(): String {
        return "Configured Features for minecraft:ores"
    }

    override fun configure(registries: RegistryWrapper.WrapperLookup, entries: Entries) {
        val oreRegistry = registries.getWrapperOrThrow(RegistryKeys.CONFIGURED_FEATURE)
        entries.addAll(oreRegistry)
        entries.registerOre(ModOres.EXAMPLE_ORE, 20, YOffset.getBottom(), YOffset.fixed(64))
    }

    private fun Entries.registerOre(oreGenEntry: ModOres.OreGenEntry, count: Int, bottom: YOffset, top: YOffset) {
        val featureRef = add(oreGenEntry.configured, oreGenEntry.feature)
        val placedFeature = PlacedFeature(featureRef, listOf(
            CountPlacementModifier.of(count),
            SquarePlacementModifier.of(),
            HeightRangePlacementModifier.uniform(bottom, top)
        ))
        add(oreGenEntry.placed, placedFeature)
    }
}
```
(Note: the Kotlin-Extension Functions `fun Entries.registerOre(...)` are just syntactic sugar and can be replaced by a Java Function like:)
```java
private static void registerOre(Entries entries, OreGenEntry oreGenEntry, ...) {
  RegistryEntry<ConfiguredFeature<?, ?>> featureRef = entries.add(...)
  // ...
}
```
